qcp/matrix.py

```diff
@@ -0,0 +1,150 @@
+        nrow,ncol = self.dim()
+        sum = self
+        for i in range(nrow):
+            for j in range(ncol):
+                sum[i][j] = self.state[i][j] + other.state[i][j]
+        return sum
+        
+
+    def __sub__(self, other: Matrix) -> Matrix:
+        #Subtraction between two matrices with same dimension
+        assert (isinstance(other, type(self)))
+        assert self.dim() == other.dim(),'Cannot add matrices with different dimensions'
+        nrow,ncol = self.dim()
+        subtraction = self.state
+        for i in range(nrow):
+            for j in range(ncol):
+                subtraction[i][j] = self.state[i][j] - other.state[i][j]
+        return subtraction
+
+    def columns(self) -> MATRIX:
+        pass
+
+    def __mul__(self, other):
+        #Multiplication
+        #Check if the dimensions of the two matrices are compatible
+        assert self.dim()[1] == other.dim()[0],'Cannot add matrices with different dimensions'
+        #If the other input is an integer, multiply directly
+        if isinstance(other, int):
+            nrow,ncol = self.dim()
+            for i in range(nrow):
+              for j in range(ncol):
+                self.state[i][j] = self.state[i][j] * other
+            return self                 
+        #Product matrix is of (nrow, ncol)
+        nrow = self.dim()[0]
+        ncol = other.dim()[1]
+        n =self.dim()[1]
+        #declare empty matrix 
+        mul = Matrix.zeros(nrow,ncol)
+        for i in range(nrow):
+            for j in range(ncol):
+                for k in range(n):
+                    mul[i][j] = mul[i][j] + self.state[i][k] * other.state[k][j]
+        return mul
+        
+    #Handle right multiplication, e.g. 5*M 
+    #Since rmul is not called if two matrices multiply, no need to worry about commutation  
+    __rmul__ = __mul__
+
+    def __str__(self) -> str:
+        pass
+
+    def conjugate(self):
+        #conjugation of matrix, the state "conjugated" is reversed
+        nrow,ncol = self.dim()
+        for i in range(nrow):
+            for j in range(ncol): 
+                if isinstance(self.state[i][j],complex):  
+                 self.state[i][j] =  self.state[i][j].real -1j*self.state[i][j].imag
+        self.conjugated = not self.conjugated                
+        return self.state
+    
+    def transpose(self):
+        #transpose of matrix
+        nrow,ncol = self.dim()
+        temp = []
+        new = []
+        for i in range(ncol):
+            for j in range(nrow):
+                temp.append(self.state[j][i])
+            new.append(temp)
+            temp = [] 
+        self.state = new       
+        return self    
+
+    def adjoint(self):
+        #dagger operation
+        self.transpose()
+        self.conjugate()
+        return self
+
+    def __repr__(self):
+        #print list instead of object address
+        return "%s" %(self.state)
```

Comment:
I believe this % notation for string formatting is a python2-ism and should be avoided (may be wrong on that...)